### PR TITLE
Multi-session merge to main

### PR DESCRIPTION
**Pull request by pedrohenrikle**
Features implemented:

- Proto file was adapted to support multi-session;
- Client can now pass any session ID they want to connect;
- Server generate a new session for every ID and don't allow two or more clients connected into the same session;
     - If the client don't pass a ID, the server will generate a new unique ID;

### CODE REVIEW

File: cmd/client/client.go [hunk 1]
```diff
@@ -16,14 +17,25 @@ import (
 	"google.golang.org/grpc/credentials"
 )
 
-var url = "http://localhost:8080/cert"
+var (
+	url       = "http://localhost:8080/cert"
```
Comment (LinkIsGrim) [inline cmd/client/client.go:21]: Move address (localhost) to flag/viper

File: cmd/client/client.go [hunk 2]
```diff
@@ -16,14 +17,25 @@ import (
...
+	sessionID *string
+)
+
+func init() {
+	id := flag.String("id", "", "Session ID")
+	flag.Parse()
+	if *id != "" {
+		sessionID = id
+	}
+}
```
Comment (LinkIsGrim) [inline cmd/client/client.go:31]: Would be a good place to use https://github.com/spf13/viper

File: cmd/client/client.go [hunk 3]
```diff
@@ -39,33 +51,44 @@ func main() {
...
 	}
 
-	// Create TLS credentials
 	creds := credentials.NewTLS(&tls.Config{RootCAs: certPool})
 
-	dial, err := grpc.NewClient(
-		"localhost:50052",
-		grpc.WithTransportCredentials(creds),
-	)
+	conn, err := grpc.NewClient("localhost:50052", grpc.WithTransportCredentials(creds))
```
Comment (LinkIsGrim) [inline cmd/client/client.go:61]: Same here

File: cmd/client/client.go [hunk 4]
```diff
@@ -39,33 +51,44 @@ func main() {
...
 		panic(err)
 	}
-	defer dial.Close()
+	defer conn.Close()
 
-	client := pb.NewCommandServiceClient(dial)
+	client := pb.NewTerminalServiceClient(conn)
+
+	var sessionRes *pb.SessionResponse
+	sessionRes, err = client.RequestSession(context.Background(), &pb.SessionRequest{Id: sessionID})
```
Comment (LinkIsGrim) [inline cmd/client/client.go:70]: What happens if a session is requested with an ID that doesn't exist? Is a new session created with that ID or is an error returned?

File: cmd/server/server.go [hunk 5]
```diff
@@ -10,93 +11,171 @@ import (
...
-	// start a bash session
-	bashSession := exec.Command("bash")
-	ptmx, err := pty.Start(bashSession)
-	if err != nil {
-		return err
-	}
-	defer func() { _ = ptmx.Close() }() // close pty when finish
+func generateSessionId() string {
+	fmt.Println(time.Now().UnixNano())
+	return fmt.Sprintf("%d", time.Now().UnixNano())
```
Comment (LinkIsGrim) [inline cmd/server/server.go:46]: use [uuid-v4](https://github.com/google/uuid) to generate. industry standard.

File: cmd/server/server.go [hunk 6]
```diff
@@ -10,93 +11,171 @@ import (
...
+			fmt.Printf("Session %s is in use.\n", sessionId)
+			return &pb.SessionResponse{
+				Id:            sessionId,
+				SessionStatus: pb.SessionStatus_IN_USE,
+			}, nil
+		} else {
+			session.InUse = true // Mark session as in use
+			fmt.Printf("Marked session %s as in use.\n", sessionId)
+		}
+	} else {
```
Comment (LinkIsGrim) [inline cmd/server/server.go:74]: Getting a bit large for a single file. I'd move the logic of this block to a "session" module and session.New() function

File: cmd/server/server.go [hunk 7]
```diff
@@ -10,93 +11,171 @@ import (
...
+	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
 )
 
 type Server struct {
-	pb.UnimplementedCommandServiceServer
+	pb.UnimplementedTerminalServiceServer
+	sessions   map[string]*BashSession
+	sessionMux sync.Mutex
```
Comment (LinkIsGrim) [inline cmd/server/server.go:29]: Good job.

File: cmd/server/server.go [hunk 8]
```diff
@@ -10,93 +11,171 @@ import (
...
+}
+
+type BashSession struct {
+	Id              string
+	TerminalCommand *exec.Cmd
+	Ptmx            *os.File
```
Comment (LinkIsGrim) [inline cmd/server/server.go:35]: I'd also move this type to a session module.

Ideally, the server should only use sessions, not implement them, but this a mostly a code organization concern.

File: cmd/server/server.go [hunk 9]
```diff
@@ -10,93 +11,171 @@ import (
...
-	ch := make(chan os.Signal, 1)
-	signal.Notify(ch, syscall.SIGWINCH)
-	go func() {
-		for range ch {
-			if err := pty.InheritSize(os.Stdin, ptmx); err != nil {
-				log.Printf("Erro ao redimensionar o PTY: %s", err)
+	s.sessionMux.Lock()
+	defer s.sessionMux.Unlock()
+
+	if session, exists := s.sessions[sessionId]; exists {
```
Comment (LinkIsGrim) [inline cmd/server/server.go:63]: Nice.

File: cmd/server/server.go [hunk 10]
```diff
@@ -10,93 +11,171 @@ import (
...
+	return &pb.SessionResponse{
+		Id:            sessionId,
+		SessionStatus: pb.SessionStatus_AVAILABLE,
+	}, nil
+}
 
-	// goroutine to send the session output to client
+func (s *Server) ExecuteCommand(stream pb.TerminalService_ExecuteCommandServer) error {
+	var bashSession *BashSession
+	var err error
```
Comment (LinkIsGrim) [inline cmd/server/server.go:121]: Don't need these declarations

File: proto/gSSH.proto [hunk 11]
```diff
@@ -4,14 +4,31 @@ package container;
...
+enum SessionStatus {
+  AVAILABLE = 0;
+  IN_USE = 1;
+  TERMINATED = 2;
+}
+
+message SessionResponse {
+  string id = 1;
+  SessionStatus sessionStatus = 2;
+}
```
Comment (LinkIsGrim) [inline proto/gSSH.proto:34]:
```suggestion
}

```
Always have a newline at EOF, you can adjust your editor/formatter settings to do this automatically.

File: cmd/server/server.go [hunk 12]
```diff
@@ -111,19 +190,23 @@ func main() {
...
 	if err != nil {
 		panic(err)
 	}
 
-	fmt.Println("Listening on :50052 with TCL/SSL...")
+	fmt.Println("Listening on :50052 with TLS/SSL...")
 
+	// HTTP server to send certificate to client
 	http.HandleFunc("/cert", func(w http.ResponseWriter, r *http.Request) { http.ServeFile(w, r, "./cert/server.crt") })
 	go http.ListenAndServe("localhost:8080", nil)
```
Comment (LinkIsGrim) [inline cmd/server/server.go:202]: Change to some other value that's relative to the gRPC port, like 50052-1

File: cmd/server/server.go [hunk 13]
```diff
@@ -111,19 +190,23 @@ func main() {
...
 	defer socket.Close()
 
-	// create the TLS credentials
 	creds, err := credentials.NewServerTLSFromFile(crt, key)
 	if err != nil {
 		panic(err)
 	}
 
-	fmt.Println("Listening on :50052 with TCL/SSL...")
+	fmt.Println("Listening on :50052 with TLS/SSL...")
```
Comment (LinkIsGrim) [inline cmd/server/server.go:198]: Use an env var or flag (or both) for the port.

File: Dockerfile
Comment (LinkIsGrim) [inline Dockerfile:1]: This looks unfinished. Where's your entrypoint? Is this the Dockerfile for the server or the client?

**Review by LinkIsGrim — COMMENTED** [review]
Might be easier to handle the flag/viper stuff in a separate PR first and then adapt this one.

Comments are mostly nitpicky. Not doing anything outright bad here, code'll just be harder to maintain later.

File: Dockerfile
Comment (pedrohenrikle) [reply Dockerfile:1]: Is definitely incomplete, as soon as I finish the main code I'll start to make this...

File: cmd/client/client.go (hunk 4 above)
Comment (pedrohenrikle) [reply cmd/client/client.go:70]: A new session is created with that ID.

**Pull request closed without merging**